File: db-async-common/src/main/java/com/github/mauricio/async/db/general/ArrayRowData.scala
/*
 * Copyright 2013 Maurício Linhares
 *
 * Maurício Linhares licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

package com.github.mauricio.async.db.general

import com.github.mauricio.async.db.RowData

class ArrayRowData(row : Int, val mapping : Map[String, Int], val columns : Array[Any]) extends RowData
{

  /**
   *
   * Returns a column value by it's position in the originating query.
   *
   * @param columnNumber
   * @return
   */
  def apply(columnNumber: Int): Any = columns(columnNumber)

  /**
   *
   * Returns a column value by it's name in the originating query.
   *
   * @param columnName
   * @return
   */
  def apply(columnName: String): Any = columns( mapping(columnName) )

  /**
   *
   * Number of this row in the query results. Counts start at 0.
   *
   * @return
   */
  def rowNumber: Int = row

  def length: Int = columns.length
}

File: db-async-common/src/main/java/com/github/jasync/sql/db/general/ArrayRowData.kt
package com.github.jasync.sql.db.general

import com.github.jasync.sql.db.RowData

class ArrayRowData(val row: Int, val mapping: Map<String, Int>, val columns: Array<Any?>) : RowData {

    private val columnsList by lazy { columns.toList() }

    override fun contains(element: Any?): Boolean = columns.contains(element)
    override fun containsAll(elements: Collection<Any?>): Boolean = columnsList.containsAll(elements)
    override fun indexOf(element: Any?): Int = columns.indexOf(element)
    override fun isEmpty(): Boolean = columns.isEmpty()
    override fun iterator(): Iterator<Any?> = columns.iterator()
    override fun lastIndexOf(element: Any?): Int = columns.lastIndexOf(element)
    override fun listIterator(): ListIterator<Any?> = columnsList.listIterator()
    override fun listIterator(index: Int): ListIterator<Any?> = columnsList.listIterator(index)
    override fun subList(fromIndex: Int, toIndex: Int): List<Any?> = columnsList.subList(fromIndex, toIndex)
    override val size: Int
        get() = columns.size

    /**
     *
     * Returns a column value by it's position in the originating query.
     *
     * @param index
     * @return
     */
    override fun get(index: Int): Any? = columns[index]

    /**
     *
     * Returns a column value by it's name in the originating query.
     *
     * @param column
     * @return
     */
    override fun get(column: String): Any? = columns[mapping.getValue(column)]

    /**
     *
     * Number of this row in the query results. Counts start at 0.
     *
     * @return
     */
    override fun rowNumber(): Int = row

    fun length(): Int = columns.size
}
